### PR TITLE
Do not fold `(x + null)` in certain cases.

### PR DESCRIPTION
**Pull request by pgavlin**
Morph attempts to fold `(x + null)` into `null` under the logic that the
resulting address is out-of-bounds by definition (in other words, adding
an offset to `null` always produces `null`). Under certain
circumstances, however, this transformation is either incorrect or
undesirable:
- If `x` contains side effects, the add should not be folded, as doing
  so incorrectly discards the side effects
- If we have called morph during CSE, the add should not be folded, as
  doing so may invalidate some of CSE's internal state (in particular,
  the removed tree may contain a subtree that is scheduled for CSE)

This change excludes the folding transformation in both of these cases.

### CODE REVIEW

**Comment by pgavlin** [comment]
@briansull @erozenfeld @dotnet/jit-contrib PTAL

Fixes VSO 283795.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -12378,7 +12378,7 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                             // Dereferencing the pointer in either case will have the
                             // same effect.
 
-                            if (!gtIsActiveCSE_Candidate(op1) && varTypeIsGC(op2->TypeGet()))
+                            if (!optValnumCSE_phase && varTypeIsGC(op2->TypeGet()) && ((op1->gtFlags & GTF_ALL_EFFECT) == 0))
```
Comment (erozenfeld) [inline src/jit/morph.cpp]: I think we could relax the condition and use GTF_SIDE_EFFECT instead of GTF_ALL_EFFECT.

**Comment by erozenfeld** [comment]
Any diffs with this change?

File: src/jit/morph.cpp (hunk 1 above)
Comment (pgavlin) [inline src/jit/morph.cpp]: IIUC we cannot: doing so will discard trees that contain volatile accesses.

**Comment by pgavlin** [comment]
@erozenfeld: not as reported with jit-diff.

File: src/jit/morph.cpp (hunk 1 above)
Comment (erozenfeld) [inline src/jit/morph.cpp]: It will discard trees that contain volatile reads but not writes, which, I think, is ok in this case.

Comment (pgavlin) [inline src/jit/morph.cpp]: I don't think that dropping any volatile accesses is legal: as per ECMA-335 I.12.6.7, "an optimizing compiler that converts CIL to native code shall not remove any volatile operation,
nor shall it coalesce multiple volatile operations into a single operation."

**Comment by pgavlin** [comment]
Interestingly enough, it appears that the legacy backend fails the clause in the new test that checks that side effects are not dropped by this optimization. The side effects are not in fact dropped, but the legacy backed does dereference the array length before evaluating the index expression when performing the bounds check. This is pretty clearly incorrect; however, I'm inclined to simply disable the test and fix the legacy backend in a separate change.

@dotnet/jit-contrib opinions?

**Comment by pgavlin** [comment]
@dotnet/jit-contrib: ping. If I don't hear any other opinions, I am going to disable the regression test for the time being and file a separate issue for follow-up.

**Comment by JosephTremoulet** [comment]
> I am going to disable the regression test for the time being and file a separate issue for follow-up

SGTM (and changes LGTM)

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x86 legacy_backend Checked Build and Test
@dotnet-bot test Windows_NT x64 Checked Jit Diff Build and Test

**Comment by pgavlin** [comment]
@mmitche: many of the failed jobs appear to be due to remoting or I/O errors. Any ideas?

**Comment by pgavlin** [comment]
I have no idea why this test is still running on x86: I've explicitly excluded it in `issues.targets` and it does not run when I invoke `runtest.cmd` locally.

Does anyone have any ideas here, and/or can somebody verify that this test does not run locally?

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by adiaaida** [comment]
Legacy backend has its own exclusion file.

**Comment by hqueue** [comment]
@pgavlin ARM also make use of legacy backend and I've observed that new test DevDiv_283795 is also falied in ARM with same message as in x86.
```
FAILED   - [   0][   4s]JIT/Regression/JitBlue/DevDiv_283795/DevDiv_283795/DevDiv_283795.sh
               BEGIN EXECUTION
               /opt/media/microsd/dotnet/unittest/coreoverlays/coreclr1114rO1.corefx1031r/corerun DevDiv_283795.exe
               System.NullReferenceException: Object reference not set to an instance of an object.
                  at C.Test(Int32 i, Int32 j, Boolean execute)
                  at C.Main()
               Expected: 100
               Actual: 101
               END EXECUTION - FAILED
```
If you don't mind, could you also exclcude this test for ARM by updating below file ?
https://github.com/dotnet/coreclr/blob/master/tests/testsUnsupportedOnARM32.txt

> I am going to disable the regression test for the time being and file a separate issue for follow-up

I will follow up future fixes for legacy backend in ARM.

**Pull request merged**